core/src/main/java/org/elasticsearch/common/ParseField.java

```diff
@@ -33,12 +33,24 @@
     private static final DeprecationLogger DEPRECATION_LOGGER = new DeprecationLogger(Loggers.getLogger(ParseField.class));
 
     private final String name;
+    private final String[] alternativeNames;
     private final String[] deprecatedNames;
```

Comment:
@rmuir sure, thanks for highlighting this. You are right, this class is quite confusing without the context of the PR. I'll add some documentation around the alternativeNames and deprecated names stuff and I'll also try to document the allReplacedWith bits too